examples/timeseries_classification_from_scratch.qmd

```diff
@@ -0,0 +1,173 @@
+series.
+
+```{r}
+x_train <- as.matrix(x_train)
+dim(x_train) <- c(dim(x_train), 1)
+x_test <- as.matrix(x_test)
+dim(x_test) <- c(dim(x_test), 1)
+```
+
+## Build a model
+
+We build a Fully Convolutional Neural Network originally proposed in
+[this paper](https://arxiv.org/abs/1611.06455).
+The implementation is based on the TF 2 version provided
+[here](https://github.com/hfawaz/dl-4-tsc/).
+The following hyperparameters (kernel_size, filters, the usage of BatchNorm) were found
+via random search using [KerasTuner](https://github.com/keras-team/keras-tuner).
+
+```{r}
+input_shape <- dim(x_train)[-1]
+input_layer <- layer_input(input_shape)
+
+output_layer <- input_layer %>%
+  # First convolutional layer
+  layer_conv_1d(filters = 64, kernel_size = 3, padding = "same") %>% 
+  layer_batch_normalization() %>%
+  layer_activation_relu() %>%
+  # Second convolutional layer
+  layer_conv_1d(filters = 64, kernel_size = 3, padding = "same") %>% 
+  layer_batch_normalization() %>%
+  layer_activation_relu() %>%
+  # Third convolutional layer
+  layer_conv_1d(filters = 64, kernel_size = 3, padding = "same") %>% 
+  layer_batch_normalization() %>%
+  layer_activation_relu() %>%
+  layer_global_average_pooling_1d() %>%
+  layer_dense(num_classes, activation = "softmax")
+
+model <- keras_model(inputs = input_layer, outputs = output_layer)
+```
+
+
+## Train the model
+
+```{r}
+epochs <- 500
+batch_size <- 32
+callbacks <- list(
+  callback_model_checkpoint("best_model.h5", save_best_only = TRUE, monitor = "val_loss"),
+  callback_reduce_lr_on_plateau(monitor = "val_loss", factor = 0.5, patience = 20, min_lr = 0.0001),
+  callback_early_stopping(monitor = "val_loss", patience = 50, verbose = 1)
+)
+
+model %>% compile(
+  optimizer = "adam",
+  loss = "sparse_categorical_crossentropy",
+  metrics = list("sparse_categorical_accuracy")
+)
+history <- model %>%
+  fit(x_train, as.matrix(y_train),
+    batch_size = batch_size,
+    epochs = epochs,
+    callbacks = callbacks,
+    validation_split = 0.2,
+    verbose = 1)
+```
+
+## Evaluate model on test data
+
+```{r}
+loaded_model <- load_model_hdf5("best_model.h5")
+result <- loaded_model %>% evaluate(x_test, as.matrix(y_test))
+
+sprintf("Test loss: %s", result[["loss"]])
+sprintf("Test accuracy: %s", result[["sparse_categorical_accuracy"]])
+```
+
+## Plot the model's training and validation loss
+
+```{r}
+plot(history)
+```
+
+TODO: Look into why accuracy is different in our implementation.
+
+We can see how the training accuracy reaches almost 0.95 after 100 epochs. However, by observing the validation accuracy we can see how the network still needs training until it reaches almost 0.97 for both the validation and the training accuracy after 200 epochs. Beyond the 200th epoch, if we continue on training, the validation accuracy will start decreasing while the training accuracy will continue on increasing: the model starts overfitting.
```

Comment:
There's huge difference in loss and accuracy compared to the original Python implementation.

Our implementation:

```
Epoch 49/500
90/90 [==============================] - 2s 17ms/step - loss: 0.6895 - sparse_categorical_accuracy: 0.5306 - val_loss: 0.6950 - val_sparse_categorical_accuracy: 0.5257
Epoch 50/500
90/90 [==============================] - 2s 17ms/step - loss: 0.6893 - sparse_categorical_accuracy: 0.5212 - val_loss: 0.6965 - val_sparse_categorical_accuracy: 0.5160
Epoch 51/500
90/90 [==============================] - 2s 17ms/step - loss: 0.6887 - sparse_categorical_accuracy: 0.5382 - val_loss: 0.6944 - val_sparse_categorical_accuracy: 0.5229
Epoch 00051: early stopping
```
Python implementation:
```
Epoch 380/500
90/90 [==============================] - 1s 6ms/step - loss: 0.0467 - sparse_categorical_accuracy: 0.9872 - val_loss: 0.0947 - val_sparse_categorical_accuracy: 0.9639
Epoch 381/500
90/90 [==============================] - 1s 6ms/step - loss: 0.0464 - sparse_categorical_accuracy: 0.9885 - val_loss: 0.0922 - val_sparse_categorical_accuracy: 0.9653
Epoch 382/500
90/90 [==============================] - 1s 6ms/step - loss: 0.0449 - sparse_categorical_accuracy: 0.9885 - val_loss: 0.0918 - val_sparse_categorical_accuracy: 0.9639
Epoch 383/500
90/90 [==============================] - 1s 6ms/step - loss: 0.0438 - sparse_categorical_accuracy: 0.9889 - val_loss: 0.0905 - val_sparse_categorical_accuracy: 0.9612
Epoch 00383: early stopping

```